File: 5.DFS/DFS.cpp
#include <bits/stdc++.h>
using namespace std;

void DFS(int node, vector<vector<int>> &adj, vector<bool> &vis) {
    vis[node] = true;
    cout << node << " ";

    for (int neigh : adj[node]) {
        if (!vis[neigh]) {
            DFS(neigh, adj, vis);
        }
    }
}

int main() {
    int n = 6;
    vector<vector<int>> adj(n);

    adj[0] = {1, 2};
    adj[1] = {0, 3, 4};
    adj[2] = {0, 4};
    adj[3] = {1};
    adj[4] = {1, 2, 5};
    adj[5] = {4};

    vector<bool> vis(n, 0);
     
    cout << "BFS starting from node 0: ";
    DFS(0, adj, vis);

    return 0;
}

File: 05. Depth First Search (DFS)/DFS.java
import java.util.*;

public class DFS {

    public static void dfs(int node, List<List<Integer>> adj, boolean[] vis) {
        vis[node] = true;
        System.out.print(node + " ");

        for (int neigh : adj.get(node)) {
            if (!vis[neigh]) {
                dfs(neigh, adj, vis);
            }
        }
    }

    public static void main(String[] args) {
        int n = 6;
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) 
            adj.add(new ArrayList<>());

        adj.get(0).addAll(Arrays.asList(1, 2));
        adj.get(1).addAll(Arrays.asList(0, 3, 4));
        adj.get(2).addAll(Arrays.asList(0, 4));
        adj.get(3).addAll(Arrays.asList(1));
        adj.get(4).addAll(Arrays.asList(1, 2, 5));
        adj.get(5).addAll(Arrays.asList(4));

        boolean[] vis = new boolean[n];

        System.out.print("DFS starting from node 0: ");
        dfs(0, adj, vis);
    }
}
